src/xbot/common/subsystems/PoseSubsystem.java

```diff
@@ -0,0 +1,184 @@
+package xbot.common.subsystems;
+
+import org.apache.log4j.Logger;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import xbot.common.command.BaseSubsystem;
+import xbot.common.controls.sensors.NavImu.ImuType;
+import xbot.common.controls.sensors.XGyro;
+import xbot.common.injection.wpi_factories.WPIFactory;
+import xbot.common.math.ContiguousHeading;
+import xbot.common.math.XYPair;
+import xbot.common.properties.BooleanProperty;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+
+@Singleton
+public class PoseSubsystem extends BaseSubsystem {
+        
+    private static Logger log = Logger.getLogger(PoseSubsystem.class);
+    public final XGyro imu;
+    
+    private final DoubleProperty leftDriveDistance;
+    private final DoubleProperty rightDriveDistance;
+    
+    private final DoubleProperty totalDistanceX;
+    private final DoubleProperty totalDistanceY;
+    
+    private ContiguousHeading currentHeading;
+    private final DoubleProperty currentHeadingProp;
+    
+    private ContiguousHeading lastImuHeading;
+    
+    // These are two common robot starting positions - kept here as convenient shorthand.
+    public static final double FACING_AWAY_FROM_DRIVERS = 90;
+    public static final double FACING_TOWARDS_DRIVERS = -90;
+    
+    private final DoubleProperty currentPitch;
+    private final DoubleProperty currentRoll;
+    
+    private final DoubleProperty inherentRioPitch;
+    private final DoubleProperty inherentRioRoll;
+    
+    private double previousLeftDistance;
+    private double previousRightDistance;
+    
+    private BooleanProperty rioRotated;
+    
+    @Inject
+    public PoseSubsystem(WPIFactory factory, XPropertyManager propManager) {
+        log.info("Creating PoseSubsystem");
+        imu = factory.getGyro(ImuType.navX);
+        
+        currentHeadingProp = propManager.createEphemeralProperty("CurrentHeading", 0.0);
+        // Right when the system is initialized, we need to have the old value be
+        // the same as the current value, to avoid any sudden changes later
+        lastImuHeading = imu.getYaw();
```

Comment:
In this instance at the subsystem level this feels like a reasonable way to me to ensure this history variable is populated before anything that uses it goes. Otherwise we'd need something like a nullable double? or something to test against before reading it and then having the initialization logic there which feels net messier to me.